File: python-backend/ai/InterviewAgent.py
"""
A simplified and robust technical interview agent with enhanced error handling.
"""

import os
import uuid
import json
import sys

def check_dependencies():
    """Check if required packages are installed"""
    missing_packages = []
    
    try:
        from dotenv import load_dotenv
        print("✓ python-dotenv found")
    except ImportError:
        missing_packages.append("python-dotenv")
    
    try:
        from langchain_groq import ChatGroq
        print("✓ langchain-groq found")
    except ImportError:
        missing_packages.append("langchain-groq")
    
    try:
        from langchain_core.messages import HumanMessage, AIMessage, SystemMessage
        print("✓ langchain-core found")
    except ImportError:
        missing_packages.append("langchain-core")
    
    if missing_packages:
        print(f"\n❌ Missing packages: {', '.join(missing_packages)}")
        print("\nInstall with:")
        for pkg in missing_packages:
            print(f"  pip install {pkg}")
        return False
    
    print("✓ All dependencies found")
    return True

def load_environment():
    """Load and validate environment variables"""
    try:
        from dotenv import load_dotenv
        load_dotenv()
        
        groq_key = os.getenv("GROQ_API_KEY")
        if not groq_key:
            print("❌ GROQ_API_KEY not found in environment")
            print("Please add GROQ_API_KEY to your .env file")
            return False
        
        print("✓ Environment variables loaded")
        return True
    except Exception as e:
        print(f"❌ Error loading environment: {e}")
        return False

class TechInterviewer:
    def __init__(self):
        self.sessions = {}
        self.max_questions = 5
        self.llm = None
        
        # Initialize LLM with error handling
        try:
            from langchain_groq import ChatGroq
            self.llm = ChatGroq(
                model="moonshotai/kimi-k2-instruct",
                temperature=0.3,
                max_retries=2
            )
            print("✓ LLM initialized successfully")
        except Exception as e:
            print(f"❌ Error initializing LLM: {e}")
            raise

    def start_interview(self, tech_stack="Python, JavaScript, React", position="Software Developer"):
        """Start a new interview session"""
        try:
            session_id = uuid.uuid4().hex[:8]
            
            self.sessions[session_id] = {
                "tech_stack": tech_stack,
                "position": position,
                "question_count": 0,
                "difficulty": "beginner",
                "conversation_history": [],
                "is_complete": False
            }
            
            first_tech = tech_stack.split(',')[0].strip()
            initial_message = f"""Hello! I'm your AI interviewer for today's {position} interview.

I see your tech stack includes: {tech_stack}

Let's start with something fundamental. Can you explain what {first_tech} is and describe one project where you've used it effectively?"""

            self.sessions[session_id]["conversation_history"].append({
                "role": "interviewer", 
                "content": initial_message
            })
            
            return session_id, initial_message
            
        except Exception as e:
            print(f"❌ Error starting interview: {e}")
            return None, f"Error: {e}"

    def process_answer(self, session_id, answer):
        """Process candidate answer and generate next question"""
        try:
            if session_id not in self.sessions:
                return "❌ Session not found! Please start a new interview."
            
            session = self.sessions[session_id]
            
            if session["is_complete"]:
                return "✅ Interview already completed! Type 'summary' for recap."
            
            # Validate answer
            if not answer or len(answer.strip()) < 3:
                return "🤔 I'd like to hear more from you. Please share your thoughts or ask for clarification if needed."
            
            # Add candidate's answer to history
            session["conversation_history"].append({
                "role": "candidate", 
                "content": answer
            })
            
            # Increment question count
            session["question_count"] += 1
            
            # Check if interview should end
            if session["question_count"] >= self.max_questions:
                session["is_complete"] = True
                return self._generate_completion_message(session_id)
            
            # Generate next question using simple prompting
            next_question = self._generate_next_question(session_id)
            
            session["conversation_history"].append({
                "role": "interviewer", 
                "content": next_question
            })
            
            return next_question
            
        except Exception as e:
            print(f"❌ Error processing answer: {e}")
            return f"Error processing your answer: {e}"

    def _generate_next_question(self, session_id):
        """Generate the next interview question"""
        try:
            session = self.sessions[session_id]
            
            # Create conversation context
            recent_conversation = ""
            for msg in session["conversation_history"][-4:]:
                recent_conversation += f"{msg['role'].title()}: {msg['content']}\n\n"
            
            # Create system prompt
            system_content = f"""You are a technical interviewer for a {session['position']} role.

            INTERVIEW CONTEXT:
            - Tech Stack: {session['tech_stack']}
            - Question Number: {session['question_count'] + 1} of {self.max_questions}
            - Current Level: {session['difficulty']}

            RECENT CONVERSATION:
            {recent_conversation}

            Your role as an interviewer:  
            1. Ask ONE question at a time and wait for candidates response
            2. Start with Basics and graudally increase difficulty
            3. Ask follow-up questions based on candidates previous response
            4. Probe deeper when answerd are imcomplete or need more clarification
            5. Cover both theorotical knowledge or practicle implementation,
            6. Ask about real world probllem scenarios and their probable solutions
            7. Be encouraging through your questionning
            8. Keep track of covered topics and explore uncovered topics


            TASK: Generate the next interview question based on:
            1. The candidate's previous response
            2. Their demonstrated skill level
            3. The tech stack focus areas

            GUIDELINES:
            - Ask ONE clear, specific question
            - Adapt questions based on demonstrated knowledge level
            - When candidate says "I don't know", offer hints or redirect to related simpler topics
            - Keep questions focused and specific
            - Maintain an empathetic, professional, and encouraging tone
            - Always reference their previous response to show you're listening


             Example One : 

            [System]
            Since you have experience with Next.js, Node.js, and PostgreSQL, let’s start with something simple —  
            Can you explain how Next.js handles server-side rendering (SSR) and how it differs from static site generation (SSG)?

            [Candidate]
            SSR renders the page on each request at runtime, SSG builds it at build time...

            [System]
            That’s correct. In SSR, rendering happens per request, while SSG pre-builds pages.  
            Follow-up: If your SSR endpoint has heavy data fetching, how would you optimize the response time?

            [Candidate]
            I’d use caching, like Redis...

            [System]
            Good. Now moving deeper — Suppose you have a high-traffic app with a PostgreSQL database. You notice slow queries.  
            Walk me through a practical approach to identify and fix the issue.

            [Candidate]
            I’d start with EXPLAIN ANALYZE...

            [System]
            Great. Since you’ve shown strong fundamentals, let’s get more advanced:  
            If your Next.js app needed real-time data updates from PostgreSQL, how would you implement it efficiently?


            
            
            
            Example Two:

            [System]
            Welcome! Let’s start with the basics:  
            In Node.js, what’s the difference between synchronous and asynchronous operations?

            [Candidate]
            Async is non-blocking, sync is blocking...

            [System]
            That’s right. Can you give me an example in code where asynchronous processing would be essential?

            [Candidate]
            Maybe file reading...

            [System]
            Good. Let’s move forward — In MongoDB, how would you design a schema for storing user posts and comments?

            [Candidate]
            I’d use embedded documents for comments...

            [System]
            That’s a reasonable choice. Why embedded instead of referencing?

            [Candidate]
            Hmm... I’m not sure.

            [System]
            That’s okay. Embedding reduces the need for joins, but can increase document size.  
            Now, suppose you need to fetch posts along with comments but also filter by recent comments. How would you query that in MongoDB?


            
            
            Example Three:

            [System]
            Hi, let’s start easy:  
            What is the difference between a class component and a functional component in React?

            [Candidate]
            I don’t know...

            [System]
            No problem — Class components are older, use `this.state`, while functional ones use hooks like `useState`.  
            Can you tell me what a React hook is?

            [Candidate]
            Is it like a function?

            [System]
            Yes, it’s a special function to manage state or lifecycle in functional components.  
            Here’s an example: `useEffect` runs code after rendering.  
            Now, can you give me a case where `useEffect` might be used?

            
            
            
            Example Four :

            [System]
            How would you configure a Kubernetes deployment to auto-scale based on CPU usage?

            [Candidate]
            I don’t know.

            [System]
            Alright, here’s a hint — Kubernetes has something called HPA. Do you know what HPA stands for?

            [Candidate]
            No.

            [System]
            It’s Horizontal Pod Autoscaler. It automatically adjusts pod count based on CPU/memory usage.  
            Let’s try something simpler: Can you tell me what a Kubernetes pod is?

            
            
            
            
            Example Five:

            [System]
            Can you explain how Spark processes data across a cluster?

            [Candidate]
            It splits data and processes it in parallel.

            [System]
            That’s a good start — Can you walk me through what happens after Spark splits the data? For example, how does it track progress or handle failures?

            [Candidate]
            Not sure.

            [System]
            No worries — Spark uses DAG scheduling, keeps track of transformations, and can recompute lost partitions from lineage.  
            Now, if one node fails mid-job, what happens?



            -----------------------------------------------------------------------


            Current interview session : Focus on {session['tech_stack']} technologies
            Interview Style : Professional, emphathatic, encouraging and thorough  

            Remember : You are evaluating technical competency, problem solving skilss and in depth understaning of chosen tech stack.

            Generate only the next question, nothing else."""

            from langchain_core.messages import SystemMessage, HumanMessage
            
            messages = [
                SystemMessage(content=system_content),
                HumanMessage(content="Generate the next interview question.")
            ]
            
            response = self.llm.invoke(messages)
            return response.content.strip()
            
        except Exception as e:
            print(f"❌ Error generating question: {e}")
            # Fallback questions based on progress
            fallback_questions = [
                f"Can you explain a key concept in {session['tech_stack'].split(',')[0].strip()}?",
                "How would you approach debugging a performance issue?",
                "Describe a challenging problem you solved recently.",
                "What best practices do you follow in your development process?",
                "How do you stay updated with new technologies?"
            ]
            return fallback_questions[min(session["question_count"], len(fallback_questions)-1)]

    def _generate_completion_message(self, session_id):
        """Generate interview completion message"""
        session = self.sessions[session_id]
        return f"""🏁 **Interview Complete!**

Thank you for participating in this {session['position']} interview!

📊 **Session Summary:**
- Questions Answered: {session['question_count']}/{self.max_questions}
- Tech Stack Covered: {session['tech_stack']}
- Final Difficulty: {session['difficulty'].title()}

Type 'summary' for detailed conversation history."""

    def get_summary(self, session_id):
        """Get detailed session summary"""
        if session_id not in self.sessions:
            return "❌ Session not found!"
        
        session = self.sessions[session_id]
        
        summary = f"""
📊 **DETAILED INTERVIEW SUMMARY**
═══════════════════════════════════════════════════════════════════

🆔 Session ID: {session_id}
📋 Position: {session['position']}
🛠️  Tech Stack: {session['tech_stack']}
❓ Questions: {session['question_count']}/{self.max_questions}
📈 Difficulty: {session['difficulty'].title()}
✅ Status: {'Complete' if session['is_complete'] else 'In Progress'}

📝 **FULL CONVERSATION:**
"""
        
        for i, msg in enumerate(session["conversation_history"], 1):
            role_emoji = "🎤" if msg["role"] == "interviewer" else "👤"
            summary += f"\n{i}. {role_emoji} {msg['role'].title()}:\n{msg['content']}\n{'-'*40}\n"
        
        return summary


def main():
    """Main function with comprehensive error handling"""
    try:
        print("🚀 Starting Prep Piper Interview Agent...")
        print("="*60)
        
        # Check dependencies
        print("\n📦 Checking dependencies...")
        if not check_dependencies():
            print("\n❌ Please install missing packages and try again.")
            return
        
        # Load environment
        print("\n🔧 Loading environment...")
        if not load_environment():
            print("\n❌ Please check your .env file and try again.")
            return
        
        # Initialize interviewer
        print("\n🤖 Initializing interviewer...")
        interviewer = TechInterviewer()
        
        print("\n🎯 Welcome to Prep Piper - Technical Interview Simulator!")
        print("This AI conducts structured technical interviews based on your tech stack.\n")
        
        # Get interview setup
        print("📝 Interview Setup:")
        tech_stack = input("Enter your tech stack (comma-separated, or press Enter for default): ").strip()
        if not tech_stack:
            tech_stack = "Python, JavaScript, React"
            print(f"Using default: {tech_stack}")
        
        position = input("Enter position (default: Software Developer): ").strip()
        if not position:
            position = "Software Developer"
        
        # Start interview
        print("\n🎬 Starting interview...")
        session_id, initial_message = interviewer.start_interview(tech_stack, position)
        
        if not session_id:
            print(f"❌ Failed to start interview: {initial_message}")
            return
        
        print("="*80)
        print(f"🎯 TECHNICAL INTERVIEW STARTED")
        print(f"📋 Position: {position}")
        print(f"🛠️  Tech Stack: {tech_stack}")
        print(f"🆔 Session ID: {session_id}")
        print("="*80)
        print(f"\n🎤 Interviewer: {initial_message}")
        
        print("\n💡 Commands: 'exit', 'summary', 'save'")
        print("-" * 80)
        
        # Main interview loop
        while True:
            print("\n" + "-" * 50)
            user_input = input("\n👤 Your Response: ").strip()
            
            if user_input.lower() == 'exit':
                print("\n🏁 Interview Ended")
                print(interviewer.get_summary(session_id))
                print("\nThank you for using Prep Piper!")
                break
                
            elif user_input.lower() == 'summary':
                print(interviewer.get_summary(session_id))
                continue
                
            elif user_input.lower() == 'save':
                # Simple save to file
                try:
                    os.makedirs("interviews", exist_ok=True)
                    filename = f"interviews/{session_id}.json"
                    with open(filename, 'w') as f:
                        json.dump(interviewer.sessions[session_id], f, indent=2)
                    print(f"✅ Session saved to {filename}")
                except Exception as e:
                    print(f"❌ Save error: {e}")
                continue
                
            elif not user_input:
                print("💭 Please provide an answer, or use 'exit', 'summary', or 'save' commands.")
                continue
            
            # Process the answer and get next question
            print("🔄 Processing your answer...")
            response = interviewer.process_answer(session_id, user_input)
            print(f"\n🎤 Interviewer: {response}")
            
    except KeyboardInterrupt:
        print("\n\n⏸️  Interview interrupted by user")
    except Exception as e:
        print(f"\n🚨 Critical error: {e}")
        print(f"Error type: {type(e).__name__}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    # Add debug info
    print("🐍 Python version:", sys.version)
    print("📂 Current directory:", os.getcwd())
    print("📄 Script file:", __file__)
    print()
    
    try:
        main()
    except Exception as e:
        print(f"\n💥 Fatal error in main(): {e}")
        import traceback
        traceback.print_exc()
        input("\nPress Enter to exit...")

File: express-backend2/src/interviewAgent.ts
/**
 * A simplified and robust technical interview agent with enhanced error handling.
 * Converted from Python to TypeScript
 * 

 */

import * as fs from 'fs';
import * as path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { config } from 'dotenv';
import * as readline from 'readline';
import  {ChatGroq}  from "@langchain/groq";
import { HumanMessage,SystemMessage } from "@langchain/core/messages";




interface ConversationMessage {
  role: 'interviewer' | 'candidate';
  content: string;
}

interface InterviewSession {
  tech_stack: string;
  position: string;
  question_count: number;
  difficulty: 'beginner' | 'intermediate' | 'advanced';
  conversation_history: ConversationMessage[];
  is_complete: boolean;
}



function loadEnvironment(): boolean {
  try {
    config();
    
    const apiKey = process.env.GROQ_API_KEY || process.env.OPENAI_API_KEY || process.env.LLM_API_KEY;
    if (!apiKey) {
      console.log("❌ API key not found in environment");
      console.log("Please add GROQ_API_KEY, OPENAI_API_KEY, or LLM_API_KEY to your .env file");
      return false;
    }
    
    console.log("✓ Environment variables loaded");
    return true;
  } catch (error) {
    console.log(`❌ Error loading environment: ${error}`);
    return false;
  }
}

class TechInterviewer {
  private sessions: { [sessionId: string]: InterviewSession } = {};
  private readonly maxQuestions: number = 5;
  private readonly llm: ChatGroq;
  
  constructor() {
    try {
      this.llm = new ChatGroq({
                model:"moonshotai/kimi-k2-instruct",
                temperature:0.3,
                maxRetries:2
     });
            console.log("✓ LLM initialized successfully");
    } catch (error) {
      console.log(`❌ Error initializing LLM: ${error}`);
      throw error;
    }
  }
  
  /**
   * Start a new interview session
   */
  startInterview(techStack = "Python, JavaScript, React", position = "Software Developer"): [string | null, string] {
    try {
      const sessionId = uuidv4().substring(0, 8);
      
      this.sessions[sessionId] = {
        tech_stack: techStack,
        position: position,
        question_count: 0,
        difficulty: 'beginner',
        conversation_history: [],
        is_complete: false
      };
      
      const firstTech = techStack.split(',')[0]?.trim() || techStack;
      const initialMessage = `Hello! I'm Prep Piper, your AI interviewer for today's ${position} interview.

I see your tech stack includes: ${techStack}

Let's start with something fundamental. Can you explain what ${firstTech} is and describe one project where you've used it effectively?`;

      this.sessions[sessionId].conversation_history.push({
        role: 'interviewer',
        content: initialMessage
      });
      
      return [sessionId, initialMessage];
      
    } catch (error) {
      console.log(`❌ Error starting interview: ${error}`);
      return [null, `Error: ${error}`];
    }
  }
  
  /**
   * Process candidate answer and generate next question
   */
  async processAnswer(sessionId: string, answer: string): Promise<string> {
    try {
      if (!(sessionId in this.sessions)) {
        return "❌ Session not found! Please start a new interview.";
      }
      
      const session: InterviewSession = this.sessions[sessionId]!;
      
      if (session?.is_complete) {
        return "✅ Interview already completed! Type 'summary' for recap.";
      }
      
      // Validate answer
      if (!answer || answer.trim().length < 3) {
        return "🤔 I'd like to hear more from you. Please share your thoughts or ask for clarification if needed.";
      }
      
      // Add candidate's answer to history
      session?.conversation_history.push({
        role: 'candidate',
        content: answer
      });
      
      // Increment question count
      session.question_count += 1;
      
      // Check if interview should end
      if (session?.question_count >= this.maxQuestions) {
        session.is_complete = true;
        return this.generateCompletionMessage(sessionId);
      }
      
      // Generate next question
      const nextQuestion = await this.generateNextQuestion(sessionId);
      
      session?.conversation_history.push({
        role: 'interviewer',
        content: nextQuestion
      });
      
      return nextQuestion;
      
    } catch (error) {
      console.log(`❌ Error processing answer: ${error}`);
      return `Error processing your answer: ${error}`;
    }
  }
  
 
  private async generateNextQuestion(sessionId: string): Promise<string> {
    try {
      const session: InterviewSession | undefined = this.sessions[sessionId];
      
      if (!session) {
        return "❌ Session not found! Please start a new interview.";
      }
      
      // Create conversation context
      const recentConversation = session.conversation_history
        .slice(-4)
        .map(msg => `${msg.role.charAt(0).toUpperCase() + msg.role.slice(1)}: ${msg.content}`)
        .join('\n\n');
      
      // Create system prompt
      const systemContent = `You are a technical interviewer for a ${session.position} role.

INTERVIEW CONTEXT:
- Tech Stack: ${session.tech_stack}
- Question Number: ${session.question_count + 1} of ${this.maxQuestions}
- Current Level: ${session.difficulty}

RECENT CONVERSATION:
${recentConversation}

Your role as an interviewer:  
1. Ask ONE question at a time and wait for candidates response
2. Start with Basics and gradually increase difficulty
3. Ask follow-up questions based on candidates previous response
4. Probe deeper when answers are incomplete or need more clarification
5. Cover both theoretical knowledge and practical implementation
6. Ask about real world problem scenarios and their probable solutions
7. Be encouraging through your questioning
8. Keep track of covered topics and explore uncovered topics

TASK: Generate the next interview question based on:
1. The candidate's previous response
2. Their demonstrated skill level
3. The tech stack focus areas

GUIDELINES:
- Ask ONE clear, specific question
- Adapt questions based on demonstrated knowledge level
- When candidate says "I don't know", offer hints or redirect to related simpler topics
- Keep questions focused and specific
- Maintain an empathetic, professional, and encouraging tone
- Always reference their previous response to show you're listening

Current interview session: Focus on ${session.tech_stack} technologies
Interview Style: Professional, empathetic, encouraging and thorough  

Remember: You are evaluating technical competency, problem solving skills and in depth understanding of chosen tech stack.

Generate only the next question, nothing else.`;

      const messages = [
        new SystemMessage(systemContent),
        new HumanMessage('Generate the next interview question.')
      ];
      
      const response = await this.llm.invoke(messages);
      let content: string;
      if (typeof response.content === 'string') {
          content = response.content;
      } else if (Array.isArray(response.content)) {
          // Extract text from MessageContentComplex array
          content = response.content
          //ts-ignore
          .map(item => {
            if (typeof item === 'string') return item
            if ('text' in item) return item.text;
            return '';
        })              
        .join('')
        .trim();
      } else {
          content = '';
      }
      
      return content  || "Could you tell me more about your experience with the technologies we discussed?";
      
    } catch (error) {
      console.log(`❌ Error generating question: ${error}`);
      return "I'm having trouble generating the next question. Could you tell me more about your experience with the technologies we discussed?";
    }
  }
  

  private  generateCompletionMessage(sessionId: string): string {
    const session:InterviewSession = this.sessions[sessionId]!;
    return `🏁 **Interview Complete!**

Thank you for participating in this ${session.position} interview!

📊 **Session Summary:**
- Questions Answered: ${session.question_count}/${this.maxQuestions}
- Tech Stack Covered: ${session.tech_stack}
- Final Difficulty: ${session.difficulty.charAt(0).toUpperCase() + session.difficulty.slice(1)}

Type 'summary' for detailed conversation history.`;
  }
  
  /**
   * Get detailed session summary
   */
  getSummary(sessionId: string): string {
    if (!(sessionId in this.sessions)) {
      return "❌ Session not found!";
    }
    
    const session:InterviewSession = this.sessions[sessionId]!;
    
    let summary = `
📊 **DETAILED INTERVIEW SUMMARY**
═══════════════════════════════════════════════════════════════════

🆔 Session ID: ${sessionId}
📋 Position: ${session.position}
🛠️  Tech Stack: ${session.tech_stack}
❓ Questions: ${session.question_count}/${this.maxQuestions}
📈 Difficulty: ${session.difficulty.charAt(0).toUpperCase() + session.difficulty.slice(1)}
✅ Status: ${session.is_complete ? 'Complete' : 'In Progress'}

📝 **FULL CONVERSATION:**
`;
    
    session.conversation_history.forEach((msg, index) => {
      const roleEmoji = msg.role === 'interviewer' ? '🎤' : '👤';
      summary += `\n${index + 1}. ${roleEmoji} ${msg.role.charAt(0).toUpperCase() + msg.role.slice(1)}:\n${msg.content}\n${'-'.repeat(40)}\n`;
    });
    
    return summary;
  }
  
  /**
   * Save session to file
   */
  async saveSession(sessionId: string): Promise<void> {
    try {
      const interviewsDir = path.join(process.cwd(), 'interviews');
      
      // Create directory if it doesn't exist
      if (!fs.existsSync(interviewsDir)) {
        fs.mkdirSync(interviewsDir, { recursive: true });
      }
      
      const filename = path.join(interviewsDir, `${sessionId}.json`);
      const sessionData = JSON.stringify(this.sessions[sessionId], null, 2);
      
      fs.writeFileSync(filename, sessionData);
      console.log(`✅ Session saved to ${filename}`);
    } catch (error) {
      console.log(`❌ Save error: ${error}`);
    }
  }
}

/**
 * Create readline interface for user input
 */
function createReadlineInterface(): readline.Interface {
  return readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
}

/**
 * Async wrapper for readline question
 */
function askQuestion(rl: readline.Interface, question: string): Promise<string> {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer);
    });
  });
}

/**
 * Main function with comprehensive error handling
 */
async function main(): Promise<void> {
  try {
    console.log("🚀 Starting Prep Piper Interview Agent...");
    console.log("=".repeat(60));
    
   
    
    // Load environment
    console.log("\n🔧 Loading environment...");
    if (!loadEnvironment()) {
      console.log("\n❌ Please check your .env file and try again.");
      return;
    }
    
    // Initialize interviewer
    console.log("\n🤖 Initializing interviewer...");
    const interviewer = new TechInterviewer();
    
    console.log("\n🎯 Welcome to Prep Piper - Technical Interview Simulator!");
    console.log("This AI conducts structured technical interviews based on your tech stack.\n");
    
    const rl = createReadlineInterface();
    
    try {
      // Get interview setup
      console.log("📝 Interview Setup:");
      let techStack = await askQuestion(rl, "Enter your tech stack (comma-separated, or press Enter for default): ");
      techStack = techStack.trim();
      if (!techStack) {
        techStack = "Python, JavaScript, React";
        console.log(`Using default: ${techStack}`);
      }
      
      let position = await askQuestion(rl, "Enter position (default: Software Developer): ");
      position = position.trim();
      if (!position) {
        position = "Software Developer";
      }
      
      // Start interview
      console.log("\n🎬 Starting interview...");
      const [sessionId, initialMessage] = interviewer.startInterview(techStack, position);
      
      if (!sessionId) {
        console.log(`❌ Failed to start interview: ${initialMessage}`);
        return;
      }
      
      console.log("=".repeat(80));
      console.log(`🎯 TECHNICAL INTERVIEW STARTED`);
      console.log(`📋 Position: ${position}`);
      console.log(`🛠️  Tech Stack: ${techStack}`);
      console.log(`🆔 Session ID: ${sessionId}`);
      console.log("=".repeat(80));
      console.log(`\n🎤 Interviewer: ${initialMessage}`);
      
      console.log("\n💡 Commands: 'exit', 'summary', 'save'");
      console.log("-".repeat(80));
      
      // Main interview loop
      while (true) {
        console.log("\n" + "-".repeat(50));
        const userInput = await askQuestion(rl, "\n👤 Your Response: ");
        const trimmedInput = userInput.trim();
        
        if (trimmedInput.toLowerCase() === 'exit') {
          console.log("\n🏁 Interview Ended");
          console.log(interviewer.getSummary(sessionId));
          console.log("\nThank you for using Prep Piper!");
          break;
        }
        
        if (trimmedInput.toLowerCase() === 'summary') {
          console.log(interviewer.getSummary(sessionId));
          continue;
        }
        
        if (trimmedInput.toLowerCase() === 'save') {
          await interviewer.saveSession(sessionId);
          continue;
        }
        
        if (!trimmedInput) {
          console.log("💭 Please provide an answer, or use 'exit', 'summary', or 'save' commands.");
          continue;
        }
        
        // Process the answer and get next question
        console.log("🔄 Processing your answer...");
        const response = await interviewer.processAnswer(sessionId, trimmedInput);
        console.log(`\n🎤 Interviewer: ${response}`);
      }
      
    } finally {
      rl.close();
    }
    
  } catch (error) {
    if (error instanceof Error && error.message === 'SIGINT') {
      console.log("\n\n⏸️  Interview interrupted by user");
    } else {
      console.log(`\n🚨 Critical error: ${error}`);
      console.log(`Error type: ${error instanceof Error ? error.constructor.name : typeof error}`);
      if (error instanceof Error) {
        console.log(error.stack);
      }
    }
  }
}

// Handle SIGINT (Ctrl+C)
process.on('SIGINT', () => {
  console.log('\n\n⏸️  Interview interrupted by user');
  process.exit(0);
});

// Main execution
if (require.main === module) {
  // Add debug info
  console.log("🟢 Node.js version:", process.version);
  console.log("📂 Current directory:", process.cwd());
  console.log("📄 Script file:", __filename);
  console.log();
  
  main().catch((error) => {
    console.log(`\n💥 Fatal error in main(): ${error}`);
    if (error instanceof Error) {
      console.log(error.stack);
    }
    process.exit(1);
  });
}